### DISCUSSION TITLE
Unclear why pytest.raises() isn't catching this exception

### DISCUSSION THREAD

**Discussion by Evan0000000000** [pytest-dev/pytest#12015] [Q&A] [answered]
Given this pared down example:
```
# fixtures.py
import pytest

@pytest.fixture
def force_oserror():
    '''
    Raise a specific kind of exception to test handling
    '''
    raise OSError()

def force_oserror_nonfixture():
    '''
    Raise a specific kind of exception to test handling
    '''
    raise OSError()

#testing.py
from fixtures import force_oserror_nofixture, force_oserror
import pytest

    def test_nonfixture(self):
        with pytest.raises(OSError):
            force_oserror_nonfixture()

    def test_fixture(self, force_oserror):
        with pytest.raises(OSError):
            force_oserror()
```

The first test passes, the second reports the OSError as if it was unexpected, despite the raises() context. 

I think the answer is that a fixture is expected to `yield` or `return` an object, right? So does that imply this error is because it's running the fixture to get the object to give to the test and getting an exception instead?

**Comment by RonnyPfannschmidt** [comment, marked as answer]
if you check carefully you will see that the error is reported at setup time, before the test even runs
the fixture is executed before the test even starts

if you want to have a fixture that provides a function, its going to look more like


```python
@pytest.fixture
def force_oserror() -> Callable[[], None]:
   return force_oserror_nonfixture
```

**Reply by Evan0000000000** [reply]
Ahh thanks Ronny I did miss that was a setup error, pretty new to writing this type of test. That makes a lot of sense, and thank you for kindly indulging a new guy question.

**Reply by Evan0000000000** [reply]
As a side follow up, if I make that `yield force_oserror_nonfixture` as the pytest docs seem to suggest for fixtures (rather than return), I get a complaint from pytest about that type hint:
```
Return type of generator function must be compatible with "Generator[Any, Any, Any]"
  Type "Generator[Any, Any, Any]" cannot be assigned to type "() -> None"Pylance[reportInvalidTypeForm](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportInvalidTypeForm)
```

That it's looking at the string `() -> None` seems curious, but I'm not clear if this is an issue in pyright (which afaik supplies the type checking to pylance) or contradiction in what pytest recommends and what pyright expects.

**Reply by RonnyPfannschmidt** [reply]
`Generator[ Callable[[], None], None, None]`

perhaps ?

btw - if there is no teardown - return is preferred
